fireant/database/base.py

```diff
@@ -26,14 +26,25 @@ class Database(object):
 
     slow_query_log_min_seconds = 15
 
-    def __init__(self, host=None, port=None, database=None, max_processes=1, max_result_set_size=200000,
-                 cache_middleware=None, concurrency_middleware=None):
+    def __init__(
+        self,
+        host=None,
+        port=None,
+        database=None,
+        max_processes=1,
+        max_result_set_size=200000,
+        middlewares=None,
+    ):
         self.host = host
         self.port = port
         self.database = database
         self.max_result_set_size = max_result_set_size
-        self.cache_middleware = cache_middleware
-        self.concurrency_middleware = concurrency_middleware or ThreadPoolConcurrencyMiddleware(max_processes)
+        self.middlewares = (
+            middlewares
+            if middlewares is not None
+            else [ThreadPoolConcurrencyMiddleware(max_processes=max_processes)]
+        )
+        self.middlewares += [log_middleware, connection_middleware]
```

Comment:
I think we're going to need a better way to apply sensible defaults in case middleware is not used to handle these cases. If the user specifies some middleware, then they might not use the right types of middleware, which is pretty difficult to check for and would be probably pretty confusing to the user who didn't know they needed to use some specific middleware.

I would say we could probably just offer `ThreadPoolConcurrencyMiddleware` as a middleware but don't use it by default, and otherwise just execute queries sequentially. The log middleware is also probably not needed by default. We should make sure the connection is still initialized if no connection middleware is used. This way, the fireant user has full control over which middlewares they want to use and what order to execute them, but there are no pitfalls if they don't use any or certain ones.

Reply:
Yes, I was also struggling a bit with the defaults. My reasoning for the choices here was purely to maintain backwards compatibility, but I guess if we indicate this properly in the version and release notes it's fine.

Reply:
after we merge this PR and my current one, we can do a major version upgrade